Less forceful preprocessing for congruence

mrhaandi opened the pull request:
This replaces `repeat intro` by the less forceful `intros` in preprocessing for `congruence`.

**Kind:** bug fix

Fixes #5394 and #13778

This naive fix *will* break ci. Is there a better way?

<!-- If there is a user-visible change in coqc/coqtop/coqchk/coq_makefile behavior and testing is not prohibitively expensive: -->
<!-- (Otherwise, remove this line.) -->
- [ ] Added / updated test-suite
- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).
- [ ] Overlay pull requests (if this breaks 3rd party developments in CI, see
https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md for details)

fajb commented:
I would propose `red;intros`. To deal with negations in the conclusion.
But not introducing regressions looks hard...

JasonGross commented:
Not sure how useful it'll be here, but we can get a sense of what sort of goals people expect congruence to solve if we @coqbot CI minimize

mrhaandi commented:
Thanks @JasonGross for the minimization comment.
I now inspected a lot of regressions and have found the following illustrating examples:
```coq
(* eager congruence : success
   lazy congruence : failure *)
Goal not (true = false).
Proof. congruence. Qed. 

Axiom P : Prop.

(* eager congruence : success
   lazy congruence : failure *)
Goal (not P) -> not P.
Proof. congruence. Qed.

Definition Q := True -> P.

(* eager congruence : failure
   lazy congruence : success *)
Goal Q -> Q.
Proof. congruence. Qed. 

(* eager congruence : success
   lazy congruence : failure *)
Goal P -> Q.
Proof. congruence. Qed.

Axiom R : Prop.

(* eager congruence : failure
   lazy congruence : success *)
Goal R -> (R = not P) -> not P.
Proof. congruence. Qed.
```

@fajb
> I would propose `red;intros`. To deal with negations in the conclusion.

This would fail for the goal `Goal Q -> Q.` (which is my main motivation, and an open bug).

I am also puzzled by the `Goal (not P) -> not P.` failure for lazy congruence while `Goal R -> (R = not P) -> not P.` succeeds.
cc @PierreCorbineau

I have the following suggestion to fix the most common `Goal not (true = false).` regression:
Something along the lines of
```coq
Ltac congruence' :=
  intros;
  (let H := fresh in match goal with |- (not ?e) => (enough (H : e -> not e) by exact (fun x => H x x)) end);
  congruence.
```
This works for every of the above goals except for `Goal P -> Q.`.
Now this one I'm not sure whether it should be in scope of `congruence` in the first place.

mrhaandi commented:
I implemented the proposed change from https://github.com/coq/coq/pull/14612#issuecomment-876326565
This fixes a good chunk of regressions.
The minimization has shown that `congruence` is often used for goals such as 
```coq
Definition Symmetric {X : Type} (r : X -> X -> Prop) :=
  forall x y, r x y -> r y x.
Goal Symmetric (fun n m : nat => n = m).
Proof. congruence. Qed.
```
This *requires* eager introduction. Since this is a crucial use-case, I added the condition that introduction/unfolding can be performed eagerly, if the goal is convertible to `forall s1..sn, t1 = t2`.
As a result, propositional goals such as `Goal Q -> Q.` from before can be addressed without eager introduction.
The proposed bugfix has as (currently only) regression `Goal P -> Q.` from before. I would argue that such a goal is incorrect use of `congruence`.
Also, I would prefer to avoid something inefficient such as `congruence := (eager congruence) + (lazy congruence)`.

Zimmi48 commented:
It would be worth adding test cases to the test-suite for the behavior you want to preserve.

mrhaandi commented:
> It would be worth adding test cases to the test-suite for the behavior you want to preserve.

Yes. Strangely, for me currently testing via `make test-suite` is broken reporting `Package coq-core.toplevel not found`.
Alternatively, `make -f Makefile.dune test-suite` exits with `Done: 25829/25832 /bin/sh: 1: cannot create success/section_poly.v.log: Permission denied`.

Zimmi48 commented:
OK, the first one is #14624 I guess. The second one I don't know. Have you tried `make -C test-suite`?

SkySkimmer commented:
>Have you tried make -C test-suite?

That will just get #14624 again

But the unit test error doesn't matter, just ignore it.

JasonGross commented:
Should we ask the bot to minimize some of the failures again, or are these exactly the ones that you expect?

mrhaandi commented:
> Should we ask the bot to minimize some of the failures again, or are these exactly the ones that you expect?

No need right now. There are still issues that I fully understand (the `not P` trick is not strong enough, if a custom definition of `not` is used). A stronger auxiliary lemma is required to make ci happier.

mrhaandi commented:
Alright, the current bug fix only has one ci issue with `ci-color`. Now I would like to @coqbot ci minimize

mrhaandi commented:
Unfortunately, minimization shows that there are two somewhat mutually exclusive goals which congruence should be solving:
1) Eagerly unfold and introduce (current congruence, used in `ci-color`)
```coq
Require Import Morphisms Program.

Goal forall (r : nat -> nat -> Prop), Proper (eq ==> eq ==> impl) r.
Proof. congruence. Qed.
```
2) Introduce, do not unfold ([#13778](https://github.com/coq/coq/issues/13778))
```coq
Axiom p : nat -> Prop.
Definition q n := True -> p n.
Goal forall n m, n = m -> q n -> q m.
Proof. Fail congruence. Admitted.
```
My current bug fix solves 2), but not 1) which is the only ci failure. Accounting for both (basically looking at both goals simultaneously) would slow down congruence. What do you think?

ppedrot commented:
I think that you should go for a compatibility flag that switches between both behaviours, and deprecate it in the next version.

Zimmi48 commented:
If the only backward-compatibility issues can be solved by adding `repeat intro;` before `congruence`, I wonder if a compatibility flag is even necessary...

mrhaandi commented:
> I think that you should go for a compatibility flag that switches between both behaviours, and deprecate it in the next version.

I now consider to redesign the bug fix to be behind an option e.g. "Congruence Symbolic" that does no unfolding whatsoever (also not in hypotheses) such that turning on the option also fixes https://github.com/coq/coq/issues/13189.
Now turning on the option would make `congruence` fail in a lot of common use-cases for congruence, but enable very fast symbolic congruence arguments.

mrhaandi commented:
> Now turning on the option would make `congruence` fail in a lot of common use-cases for congruence, but enable very fast symbolic congruence arguments.

See PR #14657

ppedrot commented:
IIUC this PR was superseded by #14657, and the (optional) introduction of a flag is deferred to another PR. So we can close this one in the meantime.
